GraphQL interface

eshelyaron opened the pull request:
This PR adds the following:
- A lexer, parser, and quasi-quotation embedding for the full syntax of GraphQL executable documents (queries, mutation requests) as defined in [the specification](https://spec.graphql.org/draft/#).
- A writer for serializing the Prolog representation of GraphQL documents (as the parser produces them) to codes or strings.
- A main interface predicate `graphql_execute_document(+URI, +Document, -Result, +Options)` for requesting a remote GraphQL endpoint located at `URI` to execute `Document`, unifying `Result` with the result.

If the API looks alright, I'll work on a more comprehensive test suite, at least for the reader/writer parts.

eshelyaron committed:
ADDED: library(http/graphql) implementing a GraphQL reader and QQ

eshelyaron committed:
ADDED: graphql_document_to_string/3, graphql_execute_document/4

eshelyaron committed:
DOC: expand documentation of the graphql/4 quasi-quotation predicate

eshelyaron committed:
DOC: Add module documentation for library(http/graphql/read)

JanWielemaker commented:
Thanks!   Looks mostly good.   I propose to put the whole thing into one file though.   It is small enough and a single file is simply easier to manage.   Some other points:

   - After making it one file, this file must be in CMakeLists.txt `HTTP_PL_FILES` to get it installed.
   - Literals in DCG's are written as `"a"`, i.e., using double quotes.  That is standard compliant.   The SWI-Prolog DCG compiler recognises strings and compiles them appropriately.  There should be no need for `[95]`, etc., just `"_"` should do.
   - ``memberchk(C, [34, 92, 47, 0'b, 0'f, 0'n, 0'r, 0't])`` can be ``memberchk(C, `\/bfnrt`)``.
   - graphql_string_escape_hex//1 probably better reuse hex (and digit) routines from library(dcg/basics).
   - Libraries shall get dependencies using autoload/2 unless this is not possible because the library is required to load the file itself (defines operators, code needed by term_expansion/2, etc.).  Such code should use use_module/2.   You can achieve this easily by removing all imports (or adding one autoload/2), use the built-in editor and run ^C^D (insert dependencies).   Might need some editing, but often gets most right.

Now need to do some cooking :smile:

eshelyaron committed:
Move all GraphQL code to one file and cleanup dependencies

eshelyaron commented:
Thanks for the swift review! 
I had a misconception that some non-default flag was required for `"a"` literals in DCGs, that's good to know.
I've addressed the above points in the latest commit https://github.com/SWI-Prolog/packages-http/pull/153/commits/8eab5291976ce42593a5fc15315d6510412f42f4.

JanWielemaker commented:
Thanks for all the updates.    Made two more changes, re-running the dependencies (I saw some calls colored indicating they still needed unspecified autoloading), removed the asserts for paths in the test files.   You find these in many places, but they are no longer needed since the move to CMake.

Finally squashed the branch.   As this is the first release, the history is irrelevant.

Note that you can use PceEmacs or gxref/0 to find remaining autoload issues.  Another option is to load the test file, disable autoloading and run the tests.
